mesa/experimental/cell_space.py

```diff
@@ -0,0 +1,204 @@
+
+    def select_random_cell(self):
+        return random.choice(self.cells)
+
+    def select_random_agent(self):
+        return random.choice(list(self.agents))
+
+    def select(self, filter_func: Optional[Callable[[Cell], bool]] = None, n=0):
+        if filter_func is None and n == 0:
+            return self
+
+        return CellCollection(
+            {
+                cell: agents
+                for cell, agents in self._cells.items()
+                if filter_func is None or filter_func(cell)
+            }
+        )
+
+
+class Space:
+    cells: dict[Coordinate, Cell] = {}
+
+    def _connect_single_cell(self, cell):
+        ...
+
+    @cached_property
+    def all_cells(self):
+        return CellCollection({cell: cell.agents for cell in self.cells.values()})
+
+    def __iter__(self):
+        return iter(self.cells.values())
+
+    def __getitem__(self, key):
+        return self.cells[key]
+
+    def move_agent(self, agent: Agent, pos: Coordinate) -> None:
+        """Move an agent from its current position to a new position."""
+        self.cells[pos].add_agent(agent)
+
+    @property
+    def empties(self) -> CellCollection:
+        return self.all_cells.select(lambda cell: cell.is_empty)
+
+    def move_to_empty(self, agent: Agent) -> None:
+        while True:
+            new_cell = self.all_cells.select_random_cell()
+            if new_cell.is_empty:
+                new_cell.add_agent(agent)
+                return
+
+        # TODO: Adjust cutoff value for performance
+        for _ in range(len(self.all_cells) // 10):
+            new_cell = self.all_cells.select_random_cell()
+            if new_cell.is_empty:
+                new_cell.add_agent(agent)
+                return
+
+        try:
+            self.empties.select_random_cell().add_agent(agent)
+        except IndexError as err:
+            raise Exception("ERROR: No empty cell found") from err
+
+
+class Grid(Space):
+    def __init__(
+        self, width: int, height: int, torus: bool = False, moore=True, capacity=1
+    ) -> None:
+        self.width = width
+        self.height = height
+        self.torus = torus
+        self.moore = moore
+        self.capacity = capacity
+        self.cells = {
+            (i, j): Cell(i, j, capacity) for j in range(width) for i in range(height)
+        }
+
+        for cell in self.all_cells:
+            self._connect_single_cell(cell)
+
+    def _connect_single_cell(self, cell):
+        i, j = cell.coordinate
+
+        # fmt: off
+        if self.moore:
+            directions = [
+                (-1, -1), (-1, 0), (-1, 1),
+                ( 0, -1),          ( 0, 1),
+                ( 1, -1), ( 1, 0), ( 1, 1),
+            ]
+        else: # Von Neumann neighborhood
+            directions = [
+                          (-1, 0),
+                ( 0, -1),          (0, 1),
+                          ( 1, 0),
+            ]
+        # fmt: on
```

Comment:
This is an absolute work of art.

(Didn't know about `# fmt: off`, I'm going to use that everywhere now. Sorry @rht)

Reply:
`black` is focused only on IT code, not tailored for scientific code. I know that folks at https://github.com/SRI-International/QC-App-Oriented-Benchmarks resist `black` because of this.

Reply:
I have to give ChatGPT credit for the idea, I just adjusted it to look nicer